onnxruntime/core/session/onnxruntime_c_api_ep.cc

```diff
@@ -0,0 +1,733 @@
+          auto added_graph_input = std::find_if(graph_inputs_including_initializers.begin(),
+                                                graph_inputs_including_initializers.end(),
+                                                [&name](const NodeArg* entry) { return entry->Name() == name; });
+
+          if (added_graph_input == graph_inputs_including_initializers.end()) {
+            if (context) {
+              auto type_proto = std::make_unique<ONNX_NAMESPACE::TypeProto>();
+              type_proto->CopyFrom(*(input->TypeAsProto()));
+              auto& n_input = top_level_graph->GetOrCreateNodeArg(name, type_proto.get());
+              context->manually_added_graph_inputs[n_input.Name()] = &n_input;
+//              LOGS_DEFAULT(VERBOSE) << "[TensorRT EP] \t" << n_input.Name() << " is added as an explicit input into the newly built graph";
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+static void SetAllGraphInputs(Graph& graph, std::unordered_map<std::string, std::unique_ptr<SubGraphContext2>>& subgraph_context_map) {
+  // If ORT TRT doesn't manully set graph input in TensorrtExecutionProvider::SetGraphOuterScopeValuesAndInputs(),
+  // Graph::Resolve() will help set graph inputs in Graph::SetGraphInputsOutputs(), so no need to set graph inputs here.
+  std::string unique_graph_name = GetUniqueGraphName(graph);
+  if (subgraph_context_map.find(unique_graph_name) == subgraph_context_map.end() ||
+      subgraph_context_map[unique_graph_name].get()->manually_added_graph_inputs.size() == 0) {
+    return;
+  }
+
+  SubGraphContext2* context = subgraph_context_map[unique_graph_name].get();
+  std::vector<const NodeArg*> graph_inputs_including_initializers;
+  std::unordered_set<std::string> graph_inputs_including_initializers_set;
+
+  for (const auto& entry : context->inputs_and_initializers) {
+    graph_inputs_including_initializers.push_back(entry.second);
+    graph_inputs_including_initializers_set.insert(entry.first);
+  }
+
+  for (const auto& entry : context->manually_added_graph_inputs) {
+    if (graph_inputs_including_initializers_set.find(entry.first) == graph_inputs_including_initializers_set.end()) {
+      graph_inputs_including_initializers.push_back(entry.second);
+      graph_inputs_including_initializers_set.insert(entry.first);
+    }
+  }
+
+  for (const auto& node_arg : graph.GetInputsIncludingInitializers()) {
+    if (graph_inputs_including_initializers_set.find(node_arg->Name()) == graph_inputs_including_initializers_set.end()) {
+      graph_inputs_including_initializers.push_back(node_arg);
+      graph_inputs_including_initializers_set.insert(node_arg->Name());
+    }
+  }
+
+  graph.SetInputs(graph_inputs_including_initializers);
+}
+
+ORT_API_STATUS_IMPL(OrtGraphApis::OrtGraph_GetSubGraph, const OrtGraphViewer* graph, const int node_num, const size_t* node_indices, _Outptr_ const OrtGraphViewer** subgraph) {
```

Comment:
Could we add a feature or additional graph api to check whether the input of the subgraph has shape info?

For TRT EP, TRT requires onnx model to have input shape provided, so that's why TRT EP check the input shape of the subgraph, if not it suggests to run symbolic_shape_infer.py first.